content/blog/2020-11-09-modern-gc/index.md

```diff
@@ -0,0 +1,80 @@
+
+The rest of the previous work are variations on the [Immix](https://dl.acm.org/doi/10.1145/1379022.1375586) originally created by the author. In the original implementation, the heap is divided up into blocks with each block subdivided into lines. The rest of it's execution is like the Mostly Copying garbage collector except that memory allocations pin a series of lines in a block compared to pinning a whole page like was previous done. The most efficient of the variations listed is the [Reference Counted Immix](http://users.cecs.anu.edu.au/~steveb/pubs/papers/rcix-oopsla-2013.pdf). This adds a reference counter for each line instead of each allocated object. The counter sums the number of references to all valid objects in the line. When this number hits zero, the whole line is collected. However, this implementation must fall back on tracing to deal with cycles.
+
+## Contributions
+
+The main contribution of this paper is the novel combination of an optimized object map, the conservative Immix garbage collector, and delayed reference counting.
+
+The object map is a representation of the heap in bitmap form. Each bit corresponds to a possible location for a valid object to be store. All bits start as 0 with an empty heap and are flipped when that bit's corresponding address is allocated to. Likewise, when an object is ready to be deallocated, the bit corresponding to that address is reverted to 0. This allows the garbage collector to quickly check if an ambiguous reference points to a valid object and discard it if it does not. Various optimizations are implemented on top of the bitmap including using hardware instruction support and object alignment.
+
+Delay reference counting can then be implemented in the conservative Immix collector by treating each ambigious reference as a possible root. Instead of incrementing a counter on each allocation, you wait to do the counting of allocations pointed to from a possible root until you need to garbage collect. The method is then as follows, get the list of ambigious references to valid objects, increment the lines for each root reference it has, free all zero count lines. Before the next round of garbage collection, decrement all of the lines that were increment this round by the amount they were incremented. In a way, this is like tracing only at the topmost objects of the heap. Then, only if one of the topmost objects dies to you iterate through it's tree decrementing the reference count of lines looking for more objects to free.
```

Comment:
"increment the lines for each root reference it has": This is confusing in a couple of ways: "increment the lines" seems to miss a noun like "counter" or something. And "for each root reference it has" might be better put "for each root reference that points to it" or something.